src/software/new_geom/util/intersections.h

```diff
@@ -0,0 +1,80 @@
+#include "software/geom/util.h"
+#include "software/new_geom/ray.h"
+#include "software/new_geom/rectangle.h"
+#include "software/new_geom/segment.h"
+
+/**
+ * Computes the intersection of two segments.
+ *
+ * @param first
+ * @param second
+ *
+ * @return  one of:
+ *          - an empty vector if no intersections
+ *          - a vector containing a single point of intersection
+ *          - a vector containing two points representing the line segment of the overlap
+ * if both segments are collinear and overlapping
+ */
+std::vector<Point> intersection(const Segment &first, const Segment &second);
+
+/**
+ * Computes the points of intersection between a rectangle and a line segment.
+ * Will contain duplicates if segment goes through rectangle corners.
+ *
+ * @param rectangle
+ * @param segment
+ *
+ * @return a vector containing the points of intersection between a rectangle and a
+ * segment
+ */
+std::vector<Point> intersection(const Rectangle &rectangle, const Segment &segment);
+
+/**
+ * Computes the points of intersection between a ray and a line segment.
+ *
+ * @param ray
+ * @param segment
+ *
+ * @return  one of:
+ *          - an empty vector if no intersections
+ *          - a vector containing a single point of intersection
+ *          - a vector containing two points representing the line segment of the overlap
+ * if the ray and segment are collinear and overlapping
+ */
+std::vector<Point> intersection(const Ray &ray, const Segment &segment);
+
+/**
+ * Computes the point of intersection between two lines.
```

Comment:
Since this is between 2 lines, can we use the `Line` class here?

Reply:
Thought about it but it seems like we would be forcing it in since this function operates on four points and is usually a helper for finding intersections between two geometric objects. 

We would need to convert from our two geometric objects -> two lines -> four points. 

When we can just do two geometric objects -> four points.

Reply:
Ok if it's being used as a more generic way of finding intersections with other objects that's fine. Can we just add another version that does take lines then and have both?

I just want to avoid what we had before where we overused these "generic" parameters rather than the actual types it was in theory operating on (which just made using it sometimes confusing)

Reply:
Added this in latest